samples/javascript/index.js

```diff
@@ -0,0 +1,253 @@
+    // Then take the part of the method name after the 'on' to get the event type.
+    getHandledEventTypesForHandlerClass(handlerClass) {
+        return Object.getOwnPropertyNames(handlerClass.prototype)
+            .filter(methodName => methodName.startsWith('on'))
+            .map(methodName => methodName.substring(2));
+    }
+
+    getAllHandledEventTypes() {
+        return new Set(this.eventHandlerClasses.reduce((eventTypes, handlerClass) => eventTypes.concat(this.getHandledEventTypesForHandlerClass(handlerClass)), []));
+    }
+
+    createBuildEventStreamUrl(buildId) {
+        const types = [...this.allHandledEventTypes].join(',');
+        return `${this.gradleEnterpriseServerUrl}/build-export/v1/build/${buildId}/events?eventTypes=${types}`;
+    }
+
+    // Creates a map of event type -> handler instance for each event type supported by one or more handlers.
+    createBuildEventHandlers(build) {
+        return this.eventHandlerClasses.reduce((handlers, handlerClass) => {
+            const addHandler = (type, handler) => handlers[type] ? handlers[type].push(handler) : handlers[type] = [handler];
+
+            const handler = new handlerClass.prototype.constructor(build);
+
+            this.getHandledEventTypesForHandlerClass(handlerClass).forEach(eventType => addHandler(eventType, handler));
+
+            if (Object.getOwnPropertyNames(handlerClass.prototype).includes('complete')) {
+                addHandler('complete', handler);
+            }
+
+            return handlers;
+        }, {});
+    }
+
+    processBuild(build) {
+        this.numBuildsInProcess++;
+        const buildEventHandlers = this.createBuildEventHandlers(build);
+        const buildEventStreamUrl = this.createBuildEventStreamUrl(build.buildId);
+
+        createServerSideEventStream(buildEventStreamUrl, {
+            oncomplete:  () => {
+                this.finishedProcessingBuild();
+
+                // Call the 'complete()' method on any handler that has it.
+                if (buildEventHandlers.complete) {
+                    buildEventHandlers.complete.forEach(handler => handler.complete());
+                }
+            },
+            eventListeners: [
+                {
+                    eventName: 'BuildEvent',
+                    eventHandler:event => {
+                        const buildEventPayload = JSON.parse(event.data);
+                        const { eventType } = buildEventPayload.type;
+
+                        if (this.allHandledEventTypes.has(eventType)) {
+                            buildEventHandlers[eventType].forEach(handler => handler[`on${eventType}`](buildEventPayload));
+                        }
+                    }
+                }
+            ],
+            retry: {
+                interval: 2000,
+                maxRetries: 100
+            }
+        });
+    }
+
+    finishedProcessingBuild() {
```

Comment:
if this is an action, I would name this `finishBuildProcess()` or maybe `finishBuildProcessing()`. If this is considered an event handler, I would name this `onBuildProcessFinish()` or the like.